src/bootloaders/raspberrypi.c

```diff
@@ -0,0 +1,420 @@
+{
+	int res;
+
+	/* Try to exchange files... */
+	res = renameat2(AT_FDCWD, oldfilename, AT_FDCWD, newfilename, RENAME_EXCHANGE);
+	if (res == 0) {
+		/* ... and remove old file. */
+		if (g_remove(oldfilename) == -1) {
+			int err = errno;
+			g_warning("Failed to remove file %s: %s", oldfilename, g_strerror(err));
+		}
+
+		return 0;
+	}
+
+	/* ... or, try to replace file if filesystem does not support exchange. */
+	if (res == -1 && errno == EINVAL)
+		res = renameat2(AT_FDCWD, oldfilename, AT_FDCWD, newfilename, 0);
+
+	return res;
+}
+
+static RaucSlot *raspberrypi_find_config_slot_by_boot_partition(RaucConfig *config, gint partition)
+{
+	g_autofree gchar *name = g_strdup_printf("%u", partition);
+	return find_config_slot_by_bootname(config, name);
+}
+
+static gboolean raspberrypi_bootloader_get(const gchar *property, guint *value, GError **error)
+{
+	g_autofree gchar *filename = NULL;
+	gboolean res = FALSE;
+	guint32 val;
+	gint fd;
+
+	g_return_val_if_fail(property, FALSE);
+	g_return_val_if_fail(value, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	filename = g_build_filename("/sys/firmware/devicetree/base/chosen/bootloader", property, NULL);
+	fd = g_open(filename, O_RDONLY);
+	if (fd < 0) {
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Failed to open file: %s", filename);
+		return FALSE;
+	}
+
+	if (read(fd, &val, sizeof(val)) != sizeof(val)) {
+		g_set_error(
+				error,
+				R_BOOTCHOOSER_ERROR,
+				R_BOOTCHOOSER_ERROR_PARSE_FAILED,
+				"Failed to read integer from file: %s", filename);
+		goto out;
+	}
+
+	*value = g_htonl(val);
+	res = TRUE;
+
+out:
+	g_close(fd, NULL);
+
+	return res;
+}
+
+static gboolean raspberrypi_bootloader_get_partition(guint *partition, GError **error)
+{
+	return raspberrypi_bootloader_get("partition", partition, error);
+}
+
+static gboolean raspberrypi_bootloader_get_tryboot(gboolean *tryboot, GError **error)
+{
+	guint value;
+
+	if (!raspberrypi_bootloader_get("tryboot", &value, error))
+		return FALSE;
+
+	*tryboot = value ? TRUE : FALSE;
+
+	return TRUE;
+}
+
+static gboolean raspberrypi_tryboot_set(gboolean enable, GError **error)
+{
+	g_autoptr(GSubprocess) sub = NULL;
+	GError *ierror = NULL;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	sub = r_subprocess_new(G_SUBPROCESS_FLAGS_NONE, &ierror, RASPBERRYPI_VCMAILBOX,
+			"0x00038064", "4", "0", enable ? "1" : "0", NULL);
+	if (!sub) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to start " RASPBERRYPI_VCMAILBOX ": ");
+		return FALSE;
+	}
+
+	if (!g_subprocess_wait_check(sub, NULL, &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to run " RASPBERRYPI_VCMAILBOX ": ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Get booted bootname */
+gchar *r_raspberrypi_get_bootname(RaucConfig *config, GError **error)
+{
+	GError *ierror;
```

Comment:
The `scan-build` failure is triggered by the missing `= NULL` here.

Reply:
Thanks, it works!